crates/provider/src/layers/batch.rs

```diff
@@ -0,0 +1,359 @@
+        }
+    }
+
+    fn block(&self) -> Option<BlockId> {
+        if let Self::Call(block, _) = self {
+            *block
+        } else {
+            None
+        }
+    }
+}
+
+/// A provider that batches multiple requests into a single request.
+///
+/// See [`BatchLayer`] for more information.
+pub struct BatchProvider<P, N: Network = Ethereum> {
+    provider: Arc<P>,
+    inner: BatchProviderInner,
+    _pd: PhantomData<N>,
+}
+
+impl<P: fmt::Debug, N: Network> fmt::Debug for BatchProvider<P, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("BatchProvider(")?;
+        self.provider.fmt(f)?;
+        f.write_str(")")
+    }
+}
+
+impl<P: Provider<N> + 'static, N: Network> BatchProvider<P, N> {
+    fn new(inner: P, layer: &BatchLayer) -> Self {
+        let inner = Arc::new(inner);
+        let tx = BatchProviderBackend::spawn(inner.clone(), layer);
+        Self { provider: inner, inner: BatchProviderInner { tx, m3a: layer.m3a }, _pd: PhantomData }
+    }
+}
+
+#[derive(Clone)]
+struct BatchProviderInner {
+    tx: mpsc::UnboundedSender<BatchProviderMessage>,
+    m3a: Address,
+}
+
+impl BatchProviderInner {
+    async fn schedule<N: Network, T: SolValue>(
+        self,
+        msg: BatchProviderMessageKind<N>,
+    ) -> TransportResult<T>
+    where
+        T: From<<T::SolType as SolType>::RustType>,
+    {
+        let (msg, rx) = BatchProviderMessage::new(msg, self.m3a);
+        self.tx.send(msg).map_err(|_| TransportErrorKind::backend_gone())?;
+        let result_bytes = rx.await.map_err(|_| TransportErrorKind::backend_gone())??;
+        let result = T::abi_decode(&result_bytes, false).map_err(TransportErrorKind::custom)?;
+        Ok(result)
+    }
+}
+
+struct BatchProviderBackend<P, N: Network = Ethereum> {
+    inner: Arc<P>,
+    m3a: Address,
+    wait: Duration,
+    rx: mpsc::UnboundedReceiver<BatchProviderMessage>,
+    pending: Vec<BatchProviderMessage>,
+    _pd: PhantomData<N>,
+}
+
+impl<P: Provider<N> + 'static, N: Network> BatchProviderBackend<P, N> {
+    fn spawn(inner: Arc<P>, layer: &BatchLayer) -> mpsc::UnboundedSender<BatchProviderMessage> {
+        let BatchLayer { m3a, wait } = *layer;
+        let (tx, rx) = mpsc::unbounded_channel();
+        let this = Self { inner, m3a, wait, rx, pending: Vec::new(), _pd: PhantomData };
+        this.run().spawn_task();
+        tx
+    }
+
+    async fn run(mut self) {
+        'outer: loop {
+            // Wait for the first message.
+            debug_assert!(self.pending.is_empty());
+            match self.rx.recv().await {
+                Some(msg) => self.process_msg(msg),
+                None => break,
+            }
+
+            // Handle messages within the wait.
+            debug_assert!(!self.pending.is_empty());
+            'inner: loop {
+                sleep(self.wait).await;
+                match self.rx.try_recv() {
```

Comment:
this only pocesses one more message before sleeping again?

we want to drain the entire channel here

Reply:
the sleep should've been outside the loop